dynamiqs/plots/misc.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import annotations
+
+import numpy as np
+from matplotlib.axes import Axes
+from matplotlib.colors import ListedColormap, LogNorm, Normalize
+
+from ..utils.tensor_types import ArrayLike, to_numpy
+from .utils import add_colorbar, fock_ticks, optax, sample_cmap
+
+__all__ = [
+    'plot_pwc_pulse',
+    'plot_fock',
+    'plot_focks',
+]
+
+
+@optax
+def plot_pwc_pulse(
+    times: ArrayLike,
+    values: ArrayLike,
+    *,
+    ax: Axes = None,
+    ycenter: bool = True,
+    real_color: str = '#0C5DA5',
+    imag_color: str = '#845B97',
+):
+    times = to_numpy(times)
+    values = to_numpy(values)
+
+    # format times and values, for example:
+    # times  = [0, 1, 2, 3] -> [0, 1, 1, 2, 2, 3, 3, 4]
+    # values = [6, 7, 8, 9] -> [6, 6, 7, 7, 8, 8, 9, 9]
+    times = times.repeat(2)[1:-1]
+    values = values.repeat(2)
+    zeros = np.zeros_like(times)
+
+    # real part
+    ax.plot(times, values.real, alpha=0.6, lw=2.0, label='real', color=real_color)
+    ax.fill_between(times, zeros, values.real, alpha=0.2, color=real_color)
+
+    # imaginary part
+    ax.plot(times, values.imag, alpha=0.6, lw=2.0, label='imag', color=imag_color)
+    ax.fill_between(times, zeros, values.imag, alpha=0.2, color=imag_color)
```

Comment:
I find its a bit strange to do it like this. Matplotlib provides a function `plt.stairs` that achieves the same thing natively. You then have to provide an array `times` of shape (n+1) with all the edges of the PWC pulse. But then there's no arbitrary choice of beginning or end times.

Reply:
> But then there's no arbitrary choice of beginning or end times.

There is no arbitrary choice here, the array of times must be of shape `(n+1)` as in `plt.stairs`, it does the exact same thing. The reason I am using a custom method is because it's hard to separately set a transparent fill color but a darker edge color with `plt.stairs`. 🤷‍♂️

Reply:
Actually there is a difference, but we can easily adapt the current code, do we want the pulse stard and end to be at value 0.0? (compare the beginning and end of the pulse)
![4fc93866-1928-4ad6-acec-2bfc7482c9a8](https://github.com/dynamiqs/dynamiqs/assets/38159029/f620c0e5-96e1-4a66-8ea6-b17f1e991126)
![184a1f9c-c102-49fb-a4ee-f9117c3cc8c7](https://github.com/dynamiqs/dynamiqs/assets/38159029/00c3d7e9-9367-4c7b-b333-3ba4d216617a)

I would say yes (contrary to my original proposition), wdyt?

Reply:
Don't know which is which, but the second one is probably best !